src/natpmp.cpp

```diff
@@ -156,76 +156,102 @@ void natpmp::start(address local_address, std::string device)
 	// if necessary
 	m_version = version_pcp;
 
-	// PCP requires reporting the source address at the application
-	// layer so the socket MUST be bound to a specific address
-	// if the caller didn't specify one then get the first suitable
-	// address from the OS
-	if (local_address.is_unspecified())
+	std::vector<ip_interface> const net = enum_net_interfaces(
+		m_socket.get_io_service(), ec);
+
+	address gateway;
+	if (!device.empty())
```

Comment:
the main difference here is that if we have specified a device name, just look up that device, use the local address of that device and then look up if there's a default route over that device.

Otherwise, look up the default route, and then find the local address for the device in the interface list (that we already queried above).

it feels more reliable than to first find (or guess) the local address and then ask to get the the default route for that address.